Nonlinear eigenvalue solver interface

fdkong opened the pull request:
The implementation of nonlinear eigenvalue solvers is in SLEPc, and we have the interface only at the moose side. Keep moose simple by avoiding having logic algebra calculations in moose.

Nonlinear eigenvalue solvers are already merged into SLEPc/master, and we test the moose interface against SLEPc/master. If it passes, we should merge this interface.

fdkong committed:
Change "ComputeFullJacobianThread" to support different kernel types

fdkong committed:
Add a coefficent to adjust the sign of the coupledForce kernel

For the eigenvalue system, we do not need the negative coupledForce kernel because
the coupledForce lives at the right side.

fdkong committed:
Add interfaces for nonlinear eigenvalue solvers

call backs are attached to matrices, and these functions will be
retrieved at SLEPc side.

Nonlinear inverse power and Newton-based method

fdkong committed:
Add slepc_version to TestHarness

This allows users to specify which versions we should test against

fdkong committed:
Test 2, Laplace eigenvalue problems

Refs #7398

fdkong committed:
Test 3, an eigenvalue problem with deficient RHS

Refs #7398

fdkong committed:
Test 4, a nonlinear laplace eigenvalue problem

Refs #7389

fdkong committed:
Test 5, a coupled eigenvalue system

Refs #7398

fdkong committed:
A more robust way to set Dirichlet boundary conditions for generalized

eigenvalue problems

Refs #7398

fdkong committed:
A flag to indicate if we want to run a test

against PETSc development version (PETSc/master)

Refs #7398

fdkong committed:
Change tests to against PETSc development version

SLEPc will be a development version if PETSc is a developmenet version

Refs #7398

moosebuild commented:
Job `Precheck:linux-gnu` on 7eaedf7 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 610ec6ae6cb9036e1b9c7e37a0908384e5f1c5d8`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9343/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9343/style.patch | git apply -v`

andrsd commented on framework/src/kernels/CoupledForce.C:
```diff
@@ -21,19 +21,20 @@ validParams<CoupledForce>()
 CoupledForce::computeQpResidual()
 {
-  return -_v[_qp] * _test[_i][_qp];
+  return _cof * -_v[_qp] * _test[_i][_qp];
```
🤢 how about to have the unary operator in front of the whole thing?

andrsd commented on framework/include/kernels/CoupledForce.h:
```diff
@@ -41,6 +41,7 @@ class CoupledForce : public Kernel
 private:
   unsigned int _v_var;
   const VariableValue & _v;
+  Real _cof;
```
`coef`

andrsd commented on framework/src/kernels/CoupledForce.C:
```diff
@@ -21,19 +21,20 @@ validParams<CoupledForce>()
   InputParameters params = validParams<Kernel>();
 
   params.addRequiredCoupledVar("v", "The coupled variable which provides the force");
+  params.addParam<Real>("cof", 1.0, "Coefficent to multiply by the coupled force term");
```
`coef`

and

`Coefficent to multiply the coupled force term by`

andrsd commented on framework/src/kernels/CoupledForce.C:
```diff
@@ -46,6 +47,6 @@ Real
 {
   if (jvar == _v_var)
-    return -_phi[_j][_qp] * _test[_i][_qp];
+    return _cof * -_phi[_j][_qp] * _test[_i][_qp];
```
Same here

andrsd commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -36,7 +42,8 @@ getSlepcValidParams()
 
-  MooseEnum eigen_solve_type("POWER ARNOLDI KRYLOVSCHUR JACOBI_DAVIDSON");
+  MooseEnum eigen_solve_type("POWER ARNOLDI KRYLOVSCHUR JACOBI_DAVIDSON "
+                             "NONLINEAR_POWER  MONOLITH_NEWTON");
```
just 1 space

moosebuild commented:
Job `Moose Documentation:linux-clang` on ee6b626 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9343/site

andrsd commented on python/TestHarness/testers/Tester.py:
```diff
@@ -31,6 +31,7 @@ def validParams():
         params.addParam('platform',      ['ALL'], "A list of platforms for which this test will run on. ('ALL', 'DARWIN', 'LINUX', 'SL', 'LION', 'ML')")
         params.addParam('compiler',      ['ALL'], "A list of compilers for which this test is valid on. ('ALL', 'GCC', 'INTEL', 'CLANG')")
         params.addParam('petsc_version', ['ALL'], "A list of petsc versions for which this test will run on, supports normal comparison operators ('<', '>', etc...)")
+        params.addParam('slepc_version', [],"A list of slepc versions for which this test will run on, supports normal comparison operators ('<', '>', etc...)")
```
Add a space after `,`

andrsd replied:
This was not fixed. I still want a space after `,` here.

andrsd commented on test/tests/problems/eigen_problem/ne.i:
```diff
@@ -0,0 +1,76 @@
+ nx = 8
+ ny = 8
+
+ uniform_refine = 0
```
Incorrect indetation,
`uniform_refine=0` is a NOOP.

andrsd commented on test/tests/problems/eigen_problem/ne_deficient_b.i:
```diff
@@ -0,0 +1,103 @@
+ ny = 8
+
+ uniform_refine = 0
+[]
```
Incorrect indentation.
`uniform_refine = 0` is a NOOP.

andrsd commented on test/tests/problems/eigen_problem/tests:
```diff
@@ -42,4 +42,20 @@
+    input = 'ne.i'
+    cli_args = 'Executioner/eigen_solve_type=NONLINEAR_POWER Outputs/file_base=nonlinear_power'
+    csvdiff = 'nonlinear_power_eigenvalues_0001.csv'
+    slepc_version = '3.7.4'
```
Do you need `>=` here?

andrsd commented on test/tests/problems/eigen_problem/tests:
```diff
@@ -42,4 +42,20 @@
+    input = 'ne.i'
+    cli_args = 'Executioner/eigen_solve_type=MONOLITH_NEWTON Outputs/file_base=monolith_newton'
+    csvdiff = 'monolith_newton_eigenvalues_0001.csv'
+    slepc_version = '3.7.4'
```
Do you need `>=` here?

andrsd commented on test/tests/problems/eigen_problem/ane.i:
```diff
@@ -0,0 +1,85 @@
+ nx = 8
+ ny = 8
+
+ uniform_refine = 0
```
Incorrect indentation,
`uniform_refine=0` is a NOOP.

andrsd commented on test/tests/problems/eigen_problem/ane.i:
```diff
@@ -0,0 +1,85 @@
+[]
+
+# 0.0287721
+# 0.02876678286685
```
What is this?

andrsd commented on test/tests/problems/eigen_problem/ne_coupled.i:
```diff
@@ -0,0 +1,136 @@
+ nx = 8
+ ny = 8
+
+ uniform_refine = 0
```
Incorrect indentation,
`uniform_refine=0` is a NOOP.

fdkong commented:
@andrsd, 

For one comment in the previous PR #9208: why `slepc_version = []` not `[ALL]`?

SLEPc is an optional package, and we do not want to do anything if users do not put `slepc=true` or `slepc_version=something` in `tests`.

if `slepc_version = [ALL]`  by default, we then have no idea to determine if users want to run slepc or not.

andrsd commented:
Also typos in commit messages:
* c7ed750 : runt -> run
* 9a17e59 : eigenvalu -> eigenvalue

moosebuild commented:
Job `Moose Documentation:linux-clang` on f818e6c wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9343/site

moosebuild commented:
Job `Moose Documentation:linux-clang` on fd18a8e wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9343/site

fdkong committed:
Do not run subdomain partitioner against PETSc-master

Because parmetis and metis generate different partition

fdkong committed:
Do not run all_print_pps.i against PETSc-master

PETSc-master has better convergence on this example. The number of
iteration is one less than that using PETSc-maint, and then the
number of residual evaluations is reduced by 1. This causes CSV diff.

We will regold this test when we update to new release of PETSc.

Refs #7398

moosebuild commented:
Job `Moose Documentation:linux-clang` on 60cb9e2 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9343/site

moosebuild commented:
Job `Test intel:linux-intel` on cd4371a : invalidated by @fdkong

Intel compiler

moosebuild commented:
Job `Moose Documentation:linux-clang` on cd4371a wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9343/site

fdkong committed:
Do not run "lots_of_variables" test against PETSc-master

This is test unstable and has an unknown MPI issue

Refs #7398

moosebuild commented:
Job `Moose Documentation:linux-clang` on 19dcf97 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9343/site

fdkong committed:
Adjust abs_zero for two tests that have a tiny

exodiff when using PETSc-master

test/tests/time_integrators/rk-2/2d-quadratic
test/tests/time_integrators/tvdrk2/2d-quadratic

Refs #7398

moosebuild commented:
Job `modules debug:linux-gnu` on 19dcf97 : invalidated by @fdkong

timeout

moosebuild commented:
Job `App tests PETSc master:linux-gnu` on 19dcf97 : invalidated by @YaqiWang

to see if rattlesnake fails again

moosebuild commented:
Job `Moose Documentation:linux-clang` on fde56be wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9343/site

fdkong commented:
@andrsd, @YaqiWang, @permcody 

Any comments on this PR? 

In this PR, we add nonlinear eigenvalue interfaces, and the actual solver is in SLEPc-master already.

permcody commented on test/src/kernels/PMassKernel.C:
```diff
@@ -0,0 +1,41 @@
+}
+
+PMassKernel::PMassKernel(const InputParameters & parameters)
+  : Kernel(parameters), _p(getParam<Real>("p") - 2.0)
```
The parameter description string says that `_p` is an exponent, but you are subtracting 2 from it here. Perhaps the description is not complete?

fdkong replied:
This example comes from Yaqi. 

We have something like |u|^{p-2} in the equation.

fdkong replied:
Look at Test 4 at http://mooseframework.org/wiki/MooseSystems/Executioners/EigenExecutioner/

fdkong replied:
We may say "_p-2" is exponent

permcody reviewed (commented):
This looks ok - to me with the exception of my one comment. I'll assign to @andrsd for merging.

permcody commented:
@fdkong - Use the "reviewers" list if you want multiple people to look at it. The "assignees" list is usually for who should merge.

fdkong commented:
@andrsd, @YaqiWang, any comment?

andrsd commented on test/tests/preconditioners/pbp/tests:
```diff
@@ -32,5 +32,9 @@
     valgrind = 'HEAVY'
     min_reported_time = 20
+    # this test has unknown issue with PETSc-master
+    # It is related with communcators in VTK
```
`related to communicator...`

andrsd commented on test/tests/problems/eigen_problem/tests:
```diff
@@ -49,6 +49,8 @@
     cli_args = 'Executioner/eigen_solve_type=NONLINEAR_POWER Outputs/file_base=nonlinear_power'
     csvdiff = 'nonlinear_power_eigenvalues_0001.csv'
     slepc_version = '>=3.7.4'
+    # we want to test against PETSc master
```
cpt. Obvious, remove this comment - you are not using this below either. The meaning is documented in test harness.

andrsd commented on test/include/kernels/PMassKernel.h:
```diff
@@ -0,0 +1,42 @@
+ * This kernel implements (v, |u|^(p-2) u)/k, where u is the variable, v is the test function
+ * and k is the eigenvalue. When p=2, this kernel is equivalent with MassEigenKernel.
+ */
+
```
I think I missed this empty line before. Please, remove it, so we are consistent in formatting.

moosebuild commented:
Job `Moose Documentation:linux-clang` on 6a9d1aa wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9343/site

moosebuild commented:
Job `Moose Documentation:linux-clang` on 61b18d6 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9343/site

moosebuild commented:
Job `modules debug SLEPc:linux-gnu-slepc` on 61b18d6 : invalidated by @fdkong

timeout

YaqiWang commented:
Had jetlag and unintentionally fall in sleep this morning. I will review it now. Sorry.

YaqiWang commented on framework/src/base/ComputeFullJacobianThread.C:
```diff
@@ -27,13 +27,15 @@
     _dg_kernels(_nl.getDGKernelWarehouse()),
     _interface_kernels(_nl.getInterfaceKernelWarehouse()),
-    _kernels(_nl.getKernelWarehouse())
+    _kernels(_nl.getKernelWarehouse()),
```
should this member variable be removed?

YaqiWang commented on framework/src/base/ComputeFullJacobianThread.C:
```diff
@@ -54,6 +57,33 @@ ComputeFullJacobianThread::~ComputeFullJacobianThread() {}
 void
 ComputeFullJacobianThread::computeJacobian()
 {
+  const KernelWarehouse * warehouse;
```
looks to me these can be moved into constructor.

fdkong replied:
I like these in computeJacobian() more.

YaqiWang replied:
I disagree. These piece of code is static and should be moved in constructor.

YaqiWang commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -247,6 +273,95 @@ slepcSetOptions(FEProblemBase & problem)
+  PetscVector<Number> X_global(x, sys.comm()), R(r, sys.comm());
+
+  // update local solution
+  X_global.swap(X_sys);
```
is this necessary? can we directly pass X_global to computeResidualType()? Also doing this you will temporarily lose the synchronization between the solution and local solution.

fdkong replied:
We can not. Ghosted solution is required.

YaqiWang replied:
Then you do not need to do swap and update, you can directly call `X_global.localize(*sys.current_local_solution.get())`. At the end, you might want to call `sys.update()` to restore the synchronization.

fdkong replied:
I actually just copy this piece of code from libMesh. I guess they might have reason to do in this way. Before I change the code, I want comments from @jwpeterson.

jwpeterson replied:
I think we should leave the `swap` in place. No values are actually swapped, just the pointers, so it's not slow if that's what you are worried about. It should be faster than `localize` which actually copies values...

YaqiWang replied:
`update` between two `swap` is doing `localize` anyway. The purpose of these three lines can be achieved by just doing `X_global.localize(*sys.current_local_solution.get())`.

YaqiWang commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -247,6 +273,95 @@ slepcSetOptions(FEProblemBase & problem)
+  PetscVector<Number> X_global(x, sys.comm());
+
+  // update local solution
+  X_global.swap(X_sys);
```
Same here as in the FormFunction.

YaqiWang commented on framework/include/utils/MooseTypes.h:
```diff
@@ -255,10 +255,12 @@ enum SolveType
+  EST_KRYLOVSCHUR,     ///< Krylov-Schur
+  EST_JACOBI_DAVIDSON, ///< Jacobi-Davidson
+  EST_NONLINEAR_POWER, ///< Nonlinear inverse power
+  EST_MONOLITH_NEWTON  ///< Newton-based eigen solver
```
just being curious, why naming as MONOLITH_NEWTON?

fdkong replied:
Do you have a better idea? The meaning means all things are coupled together.

YaqiWang replied:
ok.

YaqiWang commented on framework/src/base/NonlinearEigenSystem.C:
```diff
@@ -42,14 +43,48 @@ assemble_matrix(EquationSystems & es, const std::string & system_name)
+  PetscContainerCreate(eigen_system.comm().get(), &container);
+  PetscContainerSetPointer(container, p);
+  PetscObjectCompose((PetscObject)petsc_mat_A, "formJacobianCtx", NULL);
+  PetscObjectCompose((PetscObject)petsc_mat_A, "formJacobianCtx", (PetscObject)container);
```
The above line differs with this line only in the last argument, intended?

fdkong replied:
Yes. The first line detaches the old function if we have, and the second line attaches the  new function

YaqiWang replied:
I see.

permcody replied:
Can we use `nullptr` when you add new lines? Also, I'm ok with C-style casts in Petsc code, but in general we try to stick to C++-casts.

YaqiWang commented on test/tests/problems/eigen_problem/ne.i:
```diff
@@ -0,0 +1,74 @@
+  csv = true
+  file_base = monolith_newton
+  execute_on = 'timestep_end'
+  [./console]
```
I remember there is a short-cut parameter for outlier variable norms.

fdkong replied:
Right now, we have to have this. We should fix this, but it is not easy.

YaqiWang replied:
If this is not necessary, I suggest to remove it. Remember, people will possibly use this as a template for their own input files.

fdkong replied:
I will remove this once I figure out a way to fix related things in moose.  I have tried several times, and did not figure out yet.

YaqiWang replied:
It is still there?

YaqiWang commented on test/tests/problems/eigen_problem/ne.i:
```diff
@@ -0,0 +1,74 @@
+  type = EigenProblem
+  eigen_problem_type = gen_non_hermitian
+  n_eigen_pairs = 1
+  n_basis_vectors = 18
```
can you explain to me the two parameters here and in the next line? These parameters seem advanced and maybe we can give default values to them and users then do not have to set them in the input.

fdkong replied:
Yes, default should work. I will remove this from input files.

YaqiWang commented on test/tests/problems/eigen_problem/ne.i:
```diff
@@ -0,0 +1,74 @@
+    variable = u
+    boundary = '0 1 2 3'
+    value = 0
+    use_displaced_mesh = true
```
why this?

fdkong replied:
You originally created this example. I just want to reproduce the same result using the new implementation. I do not know why you want to test the displaced mesh.

YaqiWang replied:
I think we can just delete this line then.

YaqiWang commented on test/tests/problems/eigen_problem/ne_deficient_b.i:
```diff
@@ -0,0 +1,101 @@
+  type = EigenProblem
+  eigen_problem_type = gen_non_hermitian
+  n_eigen_pairs = 1
+  n_basis_vectors = 18
```
Same comments as before for these two params.

YaqiWang commented on test/tests/problems/eigen_problem/ne_deficient_b.i:
```diff
@@ -0,0 +1,101 @@
+  type = Steady
+  eigen_solve_type = MONOLITH_NEWTON
+  petsc_options_iname = '-eps_power_snes_mf_operator'
+  petsc_options_value = '1'
```
Is setting this option necessary? If so, can we create another solve type like MF_MONOLITH_NEWTON?

fdkong replied:
Not have to. If we have a good idea to form Jacobian, we should not use Jacobian-free.

YaqiWang replied:
In MOOSE, Jacobian-free is the default option, I suggest to add another solve type for Jacobian-free and make that the default for eigen problem.

YaqiWang replied:
No strong opinion on this. Maybe just PJFNK?

YaqiWang replied:
We need to add a new eigen_solve_type to remove this two lines.

YaqiWang replied:
Oh, I saw you did in the last commit.

YaqiWang commented on test/tests/problems/eigen_problem/ne.i:
```diff
@@ -0,0 +1,74 @@
+
+[VectorPostprocessors]
+  [./eigenvalues]
+    type = Eigenvalues
```
In our old implementation, eigenvalue is a postprocessor and thus outputted automatically in CSV, exodus. Now we have to add the postprocessor manually, right? Do you think that it is a good idea to add these postprocessors implicitly with actions when the problem is eigen?

fdkong replied:
I like the postprocessor in the input file more.

YaqiWang replied:
When we start converting Rattlesnake inputs, we will copy this lots of times. We can have Rattlesnake do the action. So this is fine to me.

YaqiWang commented on test/tests/problems/eigen_problem/ane.i:
```diff
@@ -0,0 +1,80 @@
+  type = EigenProblem
+  eigen_problem_type = gen_non_hermitian
+  n_eigen_pairs = 1
+  n_basis_vectors = 18
```
same here.

YaqiWang commented on test/tests/problems/eigen_problem/ane.i:
```diff
@@ -0,0 +1,80 @@
+[Executioner]
+  type = Steady
+  eigen_solve_type = MONOLITH_NEWTON
+  petsc_options_iname = '-eps_power_snes_mf_operator'
```
same here

YaqiWang commented on test/tests/problems/eigen_problem/ane.i:
```diff
@@ -0,0 +1,80 @@
+[]
+
+
+[VectorPostprocessors]
```
same

YaqiWang commented on test/tests/problems/eigen_problem/ane.i:
```diff
@@ -0,0 +1,80 @@
+  csv = true
+  file_base = ane
+  execute_on = 'timestep_end'
+  [./console]
```
same

YaqiWang commented on test/tests/problems/eigen_problem/ne_coupled.i:
```diff
@@ -0,0 +1,134 @@
+  type = EigenProblem
+  eigen_problem_type = gen_non_hermitian
+  n_eigen_pairs = 1
+  n_basis_vectors = 18
```
same

YaqiWang commented on test/tests/problems/eigen_problem/ne_coupled.i:
```diff
@@ -0,0 +1,134 @@
+[Executioner]
+  type = Steady
+  eigen_solve_type = MONOLITH_NEWTON
+  petsc_options_iname = '-eps_power_snes_mf_operator'
```
same

YaqiWang commented on test/tests/problems/eigen_problem/ne_coupled.i:
```diff
@@ -0,0 +1,134 @@
+  [../]
+[]
+
+[VectorPostprocessors]
```
same

YaqiWang commented on test/tests/problems/eigen_problem/ne_coupled.i:
```diff
@@ -0,0 +1,134 @@
+  csv = true
+  file_base = ne_coupled
+  execute_on = 'timestep_end'
+  [./console]
```
same

YaqiWang commented on framework/src/bcs/NodalBC.C:
```diff
@@ -45,7 +45,8 @@ NodalBC::NodalBC(const InputParameters & parameters)
     _save_in_strings(parameters.get<std::vector<AuxVariableName>>("save_in")),
-    _diag_save_in_strings(parameters.get<std::vector<AuxVariableName>>("diag_save_in"))
+    _diag_save_in_strings(parameters.get<std::vector<AuxVariableName>>("diag_save_in")),
+    _is_eigen(false)
```
I did not see `_is_eigen` is used anywhere in this BC. Am I missing something?

fdkong replied:
Good finding.

andrsd replied:
I thought that would be useful in child classes, would it not?

fdkong replied:
I missed one line code. It is useful in NodalBC

YaqiWang commented:
@fdkong I have done the review. This PR is good!

fdkong committed:
Added matrix-free versions of nonlinear eigenvalue solvers

MF_MONOLITH_NEWTON and MF_NONLINEAR_POWER

Refs #7398

YaqiWang commented on test/tests/problems/eigen_problem/ne.i:
```diff
@@ -0,0 +1,71 @@
+[Problem]
+  type = EigenProblem
+  eigen_problem_type = gen_non_hermitian
+  which_eigen_pairs = TARGET_MAGNITUDE
```
What does this mean? The smallest eigenvalue and corresponding eigen vector?

fdkong replied:
Yes. It will give us the smallest eigenvalue when using the nonlinear power and the monolithic newton

YaqiWang replied:
@fdkong sorry for being picky. This will go into rattlesnake almost everywhere, so I want to make sure the input makes sense in a user's point of view. It would make more sense here to use `SMALLEST_MAGNITUDE` and make it the default value for the power method and PJFNK eigen executioner. Another separate comment is that: can we make `gen_non_hermitian` the default `eigen_problem_type`?

fdkong replied:
The general user may not consider the generalized eigenvalue problems at all. We have `non_hertimtian` as the default. 

Nonlinear power and monolith newton are designed for rattlesnake. So how about we add `gen_non_hermitian` and `TARGET_MAGNITUDE` as the default setting for these two algorithms? 

@YaqiWang  may not care about other algorithms at all.

YaqiWang replied:
I thought more on this. These input parameters might be fine. Rattlesnake can have its own action to automate the adding of them.

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/98156/) on f29dfb0 wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/9343/site)

fdkong commented:
@YaqiWang  any more comments?

fdkong committed:
Do not compute variable norms if it is an eigenvalue prolblem.

Because we do not know to define RHS for eigenvalue problems.

Refs #7398

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/98194/) on 0a9d02d wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/9343/site)

YaqiWang commented:
I approve this PR.

fdkong commented:
How do you think? @andrsd

fdkong commented:
How do you think @permcody?

fdkong commented:
Ping @andrsd  @permcody

permcody commented on framework/src/base/NonlinearEigenSystem.C:
```diff
@@ -42,14 +43,46 @@ assemble_matrix(EquationSystems & es, const std::string & system_name)
+
+      PetscObjectCompose((PetscObject)petsc_mat_B, "formFunctionCtx", NULL);
+      PetscObjectCompose((PetscObject)petsc_mat_B, "formFunctionCtx", (PetscObject)container);
+      PetscObjectCompose((PetscObject)petsc_mat_B, "formJacobianCtx", NULL);
```
`NULL` -> `nullptr`

fdkong replied:
In PETSc, we always stay with `NULL`. But I changed to `nullptr` according to your comments in this PR.

permcody replied:
Well of course you do, `nullptr` is C++11 or above. The reason this is important is that `nullptr` is actually a pointer type whereas, `NULL` is a typedef to 0. In C++11 where you have things like `auto` and initializer lists, having the correct type is important. There are places where NULL can cause issues because the LHS of an expression is expecting to receive the type info from the RHS (note that this never is an issue in C or C++03).

permcody commented on framework/src/outputs/Console.C:
```diff
@@ -485,6 +486,11 @@ Console::writeVariableNorms()
 
+  // if it is an eigenvalue prolblem, we do not know to define RHS,
+  // and then we do not know how to compute variable norms
+  if (dynamic_cast<EigenProblem *>(_problem_ptr) != 0)
```
0 -> `nullptr`

permcody reviewed (changes requested):
Just a couple of minor comments

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/98621/) on 4297b3f wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/9343/site)
